clang/test/SemaSYCL/free_function_kernel_params.cpp

```diff
@@ -55,3 +55,119 @@ template void ff_3(int* ptr, int start, int end);
+// CHECK-NEXT: UnaryOperator {{.*}} 'Derived' lvalue prefix '*' cannot overflow
+// CHECK-NEXT: CXXReinterpretCastExpr {{.*}} 'Derived *' reinterpret_cast<Derived *> <BitCast>
+// CHECK-NEXT: UnaryOperator {{.*}} '__generated_Derived *' prefix '&' cannot overflow
+// CHECK-NEXT: DeclRefExpr {{.*}} '__generated_Derived' lvalue ParmVar {{.*}} '__arg_S2' '__generated_Derived'
+// CHECK-NEXT: CXXConstructExpr {{.*}} 'Derived1' 'void (const Derived1 &) noexcept'
+// CHECK-NEXT: ImplicitCastExpr {{.*}} 'const Derived1' lvalue <NoOp>
+// CHECK-NEXT: UnaryOperator {{.*}} 'Derived1' lvalue prefix '*' cannot overflow
+// CHECK-NEXT: CXXReinterpretCastExpr {{.*}} 'Derived1 *' reinterpret_cast<Derived1 *> <BitCast>
+// CHECK-NEXT: UnaryOperator {{.*}} '__generated_Derived1 *' prefix '&' cannot overflow
+// CHECK-NEXT: DeclRefExpr {{.*}} '__generated_Derived1' lvalue ParmVar {{.*}} '__arg_S3' '__generated_Derived1'
+
+template <typename T1, typename T2>
+__attribute__((sycl_device))
+[[__sycl_detail__::add_ir_attributes_function("sycl-single-task-kernel", 0)]]
+  void ff_6(T1 S1, T2 S2, int end) {
+}
+
+// Explicit instantiation.
+template void ff_6(Agg S1, Derived1 S2, int);
+// CHECK: FunctionDecl {{.*}}__sycl_kernel{{.*}}'void (__generated_Agg, __generated_Derived1, int)'
+// CHECK-NEXT: ParmVarDecl {{.*}} used __arg_S1 '__generated_Agg'
+// CHECK-NEXT: ParmVarDecl {{.*}} used __arg_S2 '__generated_Derived1'
+// CHECK-NEXT: ParmVarDecl {{.*}} used __arg_end 'int'
+// CHECK-NEXT: CompoundStmt
+// CHECK-NEXT: CallExpr {{.*}} 'void'
+// CHECK-NEXT: ImplicitCastExpr {{.*}} 'void (*)(Agg, Derived1, int)' <FunctionToPointerDecay>
+// CHECK-NEXT: DeclRefExpr {{.*}} 'void (Agg, Derived1, int)' lvalue Function {{.*}} 'ff_6' 'void (Agg, Derived1, int)'
+// CHECK-NEXT: CXXConstructExpr {{.*}} 'Agg' 'void (const Agg &) noexcept'
+// CHECK-NEXT: ImplicitCastExpr {{.*}} 'const Agg' lvalue <NoOp>
+// CHECK-NEXT: UnaryOperator {{.*}} 'Agg' lvalue prefix '*' cannot overflow
+// CHECK-NEXT: CXXReinterpretCastExpr {{.*}} 'Agg *' reinterpret_cast<struct Agg *> <BitCast>
+// CHECK-NEXT: UnaryOperator {{.*}} '__generated_Agg *' prefix '&' cannot overflow
+// CHECK-NEXT: DeclRefExpr {{.*}} '__generated_Agg' lvalue ParmVar {{.*}} '__arg_S1' '__generated_Agg'
+// CHECK-NEXT: CXXConstructExpr {{.*}} 'Derived1' 'void (const Derived1 &) noexcept'
+// CHECK-NEXT: ImplicitCastExpr {{.*}} 'const Derived1' lvalue <NoOp>
+// CHECK-NEXT: UnaryOperator {{.*}} 'Derived1' lvalue prefix '*' cannot overflow
+// CHECK-NEXT: CXXReinterpretCastExpr {{.*}} 'Derived1 *' reinterpret_cast<class Derived1 *> <BitCast>
+// CHECK-NEXT: UnaryOperator {{.*}} '__generated_Derived1 *' prefix '&' cannot overflow
+// CHECK-NEXT: DeclRefExpr {{.*}} '__generated_Derived1' lvalue ParmVar {{.*}} '__arg_S2' '__generated_Derived1'
+// CHECK-NEXT: ImplicitCastExpr {{.*}} 'int' <LValueToRValue>
+// CHECK-NEXT: DeclRefExpr {{.*}} 'int' lvalue ParmVar {{.*}} '__arg_end' 'int'
```

Comment:
This test doesn't really validate code gen. Perhaps it should be in `clang/test/AST`; see the `ast-dump-*` tests there. Perhaps we should create a `clang/test/ASTSYCL` directory.

Reply:
I think this test is living in the right folder. We have similar AST checks in `SemaSYCL`.

>This test doesn't really validate code gen

Can you elaborate on what you mean by this? Why should a test in `SemaSYCL` validate code gen? We have `CodeGenSYCL` for that

Reply:
Yes, I can elaborate! I apparently looked at the wrong file name and/or suffered a hallucination :)

In general though, I would expect to find tests that validate the AST under `clang/test/AST*`, tests that validate diagnostics (or lack there of) under `clang/test/Sema*`, and tests that validate IR under `clang/test/CodeGen*`. I would admit no surprise to being informed that we are not so disciplined in distributing tests in that way though! :)

Reply:
Yea we haven't separated AST tests from diagnostic tests.  They all live in `SemaSYCL`. Some tests in `SemaSYCL` which purely check AST have a `-ast` appended to the name but we aren't consistent with that either.

Reply:
This PR follows existing style. If I add just one test to a new directory, this will create a confusion for other contributors on where to add such tests. To avoid the confusion, I think we should move all AST checking tests to a separate directory in a separate PR.

Reply:
A separate PR for moving the AST tests makes sense to me.